db/hints/manager.cc

```diff
@@ -351,18 +351,15 @@ future<db::commitlog> manager::end_point_hints_manager::add_store() noexcept {
 
             return commitlog::create_commitlog(std::move(cfg)).then([this] (commitlog l) {
```

Comment:
The premise in the changelog says:

> Endpoint manager re-creates its commitlog instance every 10 seconds in
order to make sure that 1) each hint is persisted to disk no longer
than 10s after its creation, 2) each file containing hints does not
contain more than 10s worth of hints.

But, point 1 is addressed by commitlog periodic mode that flushes every 10 seconds. Point 2 seems to me to be arbitrary and unnecessary. What's the motivation behind it?

I feel like we're adding an unnecessary hierarchy here, and commitlog should be sufficient.

(I'm okay with fixing it later, if we agree that a fix is in order)

Reply:
> The premise in the changelog says:
> 
> > Endpoint manager re-creates its commitlog instance every 10 seconds in
> > order to make sure that 1) each hint is persisted to disk no longer
> > than 10s after its creation, 2) each file containing hints does not
> > contain more than 10s worth of hints.
> 
> But, point 1 is addressed by commitlog periodic mode that flushes every 10 seconds. Point 2 seems to me to be arbitrary and unnecessary. What's the motivation behind it?

The hints manager takes care not to replay hints which are older than gc_grace_seconds. Hint creation time is not stored in the commitlog, so instead it is approximated by commitlog segment modification time minus 10s - which is a lower bound for the hint creation time.

Reply:
The mutations include timestamps, and we could check the timestamps. Creating a file every 10 seconds just for that is excessive. Note that creating files comes with its own cost in disk ops and fragmentation.

Reply:
@avikivity Timestamp on the mutation has nothing to do with the gc_grace related logic in question. These are two different timestamps: data time stamp for the former and the mutation (wall clock) write time in the later. And the later is not included in the mutation.

To remind you we've been discussing this at least once during the last year and at least 3 times during the life span of the HH code.

The reason why this logic exists is because @gleb-cloudius and you were shocked by the size of the patch that was required in order to allow extension of the serialization format of the commitlog (today it's hard coded) back when HH code was introduced. My initial version of the HH code had this.

So, in order to avoid this change this approach has been suggested and implemented.